### PR TITLE
Add Automated Push Notifications for Shifts and Posts

### PR DESCRIPTION
**Pull request by mikelxc**
From Shamikh: 
Implemented Push Notifications for Shifts and Posts
1. Testing Notifications for Shifts
- Run ```yarn flush``` and ```yarn dev:start``` and make your own team 
- Enabling Notifications on Web: Navigate to http://localhost:5000/notif and click "Allow Notifications" in the browser pop up (rest assured this set up is temporary!) 
- Enabling Notifications on Mobile: Navigate to Notifications tab on app and select "Allow Notifications" in the pop up (waiting on updated build for mobile to be available). 
- Go to [http://localhost:5000/app/calendar](http://localhost:5000/app/calendar) and make a shift for yourself on the current day at a close future time 
- Verify that you receive a Notification about your shift 10 minutes before the shift's start time. 
2. Testing Notifications for Posts (Line Monitor Announcements) 
- Go to [http://localhost:5000/admin/posts](http://localhost:5000/admin/posts) and make a new post. 
- All subscribed users should receive a notification.
3. Testing Notifications Endpoint (Used to make our own Notifications History/Inbox)
- Be logged into GTHC with Shibboleth
- Navigate to [http://localhost:5000/api/v1/notifications/](http://localhost:5000/api/v1/notifications/) to get all (past and future) notifications for the current logged on user. The shift notifications can be filtered using their start_time attributes to get a notification history. 
- Use http://localhost:5000/api/v1/notifications/<user_id> to get the notifications user with that user_id. The shift notifications are under "shift_notifications" in the JSON, and Posts/Line Monitor announcements are under "announcements". 
Let Shamikh know if you encounter any issues during testing or if you don't get a notification. Test consciously and please be wary of spamming the OneSignal API with requests. All requests are logged in the server logs. If you see anything suspicious in the logs, please let me know. Thanks so much! :) 

From Mike: 
Added OneSignal integration 
Upgraded all the dependencies to latest versions, removed unnecessary packages to improve compiling speed. 
Migrated to rails webpacker native babel to solve conflicts.

### CODE REVIEW

**Comment by amanmibra** [comment]
Thank you Mike, I will check this out tomorrow 👍

File: app/javascript/packs/hello_react.jsx [hunk 1]
```diff
@@ -2,25 +2,25 @@
...
-//   name: PropTypes.string
-// }
-//
-// document.addEventListener('DOMContentLoaded', () => {
-//   ReactDOM.render(
-//     <Hello name="React" />,
-//     document.body.appendChild(document.createElement('div')),
-//   )
-// })
+import React from 'react'
```
Comment (amanmibra) [inline app/javascript/packs/hello_react.jsx]: This fileis generated from webpacker iirc, we should delete it.

Comment (amanmibra) [reply app/javascript/packs/hello_react.jsx]: This probably is what caused your rendering errors

Comment (amanmibra) [reply app/javascript/packs/hello_react.jsx]: As it conflicts with kville.jsx

**Review by amanmibra — COMMENTED** [review]
Just did a quick look at files changed. While you wait for my review, I would go ahead and test the website locally as if you are user and try playing around with everything and see if anything breaks.

**Comment by amanmibra** [comment]
@mikelxc Please make the changes I mentioned above

**Comment by amanmibra** [comment]
@shamystic lmk when this is ready for review

File: .idea/misc.xml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (amanmibra) [inline .idea/misc.xml]: Remove these `.xml` files

**Comment by shamystic** [comment]
Thank you for taking a look at this PR! Here's how you can go about testing: 
1. Testing Notifications for Shifts
- Run ```yarn flush``` and ```yarn dev:start``` and make your own team 
- Enabling Notifications on Web: Navigate to http://localhost:5000/notif and click "Allow Notifications" in the browser pop up (rest assured this set up is temporary!) 
- Enabling Notifications on Mobile: Navigate to Notifications tab on app and select "Allow Notifications" in the pop up (waiting on updated build for mobile to be available). 
- Go to [http://localhost:5000/app/calendar](http://localhost:5000/app/calendar) and make a shift for yourself on the current day at a close future time 
- Verify that you receive a Notification about your shift 10 minutes before the shift's start time. 
2. Testing Notifications for Posts (Line Monitor Announcements) 
- Go to [http://localhost:5000/admin/posts](http://localhost:5000/admin/posts) and make a new post. 
- All subscribed users should receive a notification (don't use this test more than once)
3. Testing Notifications Endpoint (Used to make our own Notifications History/Inbox)
- Be logged into GTHC with Shibboleth
- Navigate to [http://localhost:5000/api/v1/notifications/](http://localhost:5000/api/v1/notifications/) to get all (past and future) notifications for the current logged on user. The shift notifications can be filtered using their start_time attributes to get a notification history. 
- Use http://localhost:5000/api/v1/notifications/<user_id> to get the notifications user with that user_id. The shift notifications are under "shift_notifications" in the JSON, and Posts/Line Monitor announcements are under "announcements". 
Let Shamikh know if you encounter any issues during testing or if you don't get a notification. Test consciously and please be wary of spamming the OneSignal API with requests. All requests are logged in the server logs. If you see anything suspicious in the logs, please let me know. Thanks so much! :)

File: README.md [hunk 3]
```diff
@@ -4,6 +4,30 @@
 
 # README
 
+## Testing Notifications
```
Comment (amanmibra) [inline README.md]: Remove all changes to this file.

File: app/controllers/pages_controller.rb [hunk 4]
```diff
@@ -1,4 +1,19 @@
...
+      @current_user = User.find_or_create_by_oauth(user_info)
+      @current_user
+    else
+      @current_user ||= session[:user_id] && User.find_by(id: session[:user_id])
+      @current_user
+    end
+  end
+
+  helper_method :current_user
+
```
Comment (amanmibra) [inline app/controllers/pages_controller.rb:18]: Why did you add this? There is already a `current_user` function. Let's much sure we are always following `D.R.Y.`

File: app/models/user.rb [hunk 5]
```diff
@@ -11,19 +11,20 @@ class User < ApplicationRecord
...
-    email = user_info["email"]
+    netid = user_info['dukeNetID']
+    name = user_info['name']
+    email = user_info['email']
 
     # make sure User does not throw any index errors
-    if email == "" || email == nil
-      email = netid + "@duke.edu"
+    if email == '' || email == nil
+      email = netid + '@duke.edu'
```
Comment (amanmibra) [inline app/models/user.rb]: We use double quotations everywhere else, so, let's not change some of them from double to single.

File: app/views/pages/notif.html.erb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+<script>
+  var OneSignal = window.OneSignal || [];
+  OneSignal.push(function() {
+    OneSignal.init({
+      appId: "b290fd9a-eedf-44b0-8bfd-6a37646957b6",
+    });
+  });
+
+  let netid = '<%= current_user.netid %>'; 
+  console.log(netid)
```
Comment (amanmibra) [inline app/views/pages/notif.html.erb]: Just a note, always make sure all print and log statements are removed before adding to staging

File: db/migrate/20200103010349_change_notification_id_to_be_string_in_notifications.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class ChangeNotificationIdToBeStringInNotifications < ActiveRecord::Migration[5.2]
```
Comment (amanmibra) [inline db/migrate/20200103010349_change_notification_id_to_be_string_in_notifications.rb:1]: **Tip:** It looks like to me you create a migration for every single attribute to the Notification model. You can actually do this all in one migration.

File: public/OneSignalSDKUpdaterWorker.js [hunk 8]
```diff
@@ -0,0 +1 @@
+importScripts('https://cdn.onesignal.com/sdks/OneSignalSDKWorker.js');
```
Comment (amanmibra) [inline public/OneSignalSDKUpdaterWorker.js]: Go ahead and remove all of these js files by Mike.

File: test/controllers/notifications_controller_test.rb [hunk 9]
```diff
@@ -0,0 +1,48 @@
+require 'test_helper'
+
+class NotificationsControllerTest < ActionDispatch::IntegrationTest
```
Comment (amanmibra) [inline test/controllers/notifications_controller_test.rb:3]: Nice job adding a test file, we need to do this more often.

File: app/views/pages/notif.html.erb (hunk 6 above)
Comment (amanmibra) [reply app/views/pages/notif.html.erb]: Actually this whole file should be removed. This will be done via React.

File: app/javascript/packs/hello_react.jsx [hunk 10]
```diff
@@ -2,25 +2,25 @@
 // like app/views/layouts/application.html.erb. All it does is render <div>Hello React</div> at the bottom
```
Comment (amanmibra) [inline app/javascript/packs/hello_react.jsx]: Please remove this file, I asked Mike and he never did.

File: app/helpers/application_helper.rb [hunk 11]
```diff
@@ -1,2 +1,116 @@
...
+									'headings' => {'en' => title}, 'contents' => {'en' => content},
+									'included_segments' => ['Test Users']}
+			end
+			params['send_after'] = time if time
+			puts "Sending POST request to OneSignal with parameters: #{params}"
+			return params.to_json if test
+			uri = URI.parse('https://onesignal.com/api/v1/notifications')
+			http = Net::HTTP.new(uri.host, uri.port)
+			http.use_ssl = true
+			request = Net::HTTP::Post.new(uri.path,	'Content-Type'  => 'application/json;charset=utf-8', 'Authorization' => 'Basic NDY3ZjU0NTktZTUwNy00ODQyLWFmNTMtN2IzYjAyZjI5MGYx')
```
Comment (amanmibra) [inline app/helpers/application_helper.rb]: Oof, no keys please in commit. This is now in our commit history, and if someone with malicious intent really wanted to, they can find it and use our key. Fixing the commit history can only be done by either remove a commit from history manually (this prob wont work with how many commits you have), squash and merging this PR, **OR** like I suggest in other places, moving to another PR and branch and removing this branch.

Comment (amanmibra) [reply app/helpers/application_helper.rb]: The solution to having API keys is environment variables. Add the keys to the `.env` file, and let everyone on the devs team know the new key, so they can update their .env file.

File: app/helpers/application_helper.rb [hunk 12]
```diff
@@ -1,2 +1,116 @@
...
+		# @param [Array] recipients
+    # @param [String] recipient_type, one of (netids, all, players)
+		# @param [String] title of notification
+		# @param [String] content of notification
+    # @param [Boolean] test parameters without hitting OneSignal API.
+		# @return [String] OneSignal ID of scheduled notification
+		def create_notification(recipients, recipient_type, title, content, time,
+														test)
+			if recipient_type == 'netids'
+				params = {'app_id' => 'b290fd9a-eedf-44b0-8bfd-6a37646957b6',
```
Comment (amanmibra) [inline app/helpers/application_helper.rb]: More keys and priv data that should not be committed. Add to environment vars, also, I am assuming you have different app ids for dev, staging, and production, correct?

File: app/helpers/application_helper.rb [hunk 13]
```diff
@@ -1,2 +1,116 @@
...
+									'headings' => {'en' => title}, 'contents' => {'en' => content},
+									 'included_segments' => ['All']}
+			elsif recipient_type == 'players'
+				params = {'app_id' => 'b290fd9a-eedf-44b0-8bfd-6a37646957b6',
+									'headings' => {'en' => title}, 'contents' => {'en' => content},
+									'include_player_ids' => recipients}
+			else # assume sending to test users
+				params = {'app_id' => 'b290fd9a-eedf-44b0-8bfd-6a37646957b6',
+									'headings' => {'en' => title}, 'contents' => {'en' => content},
+									'included_segments' => ['Test Users']}
```
Comment (amanmibra) [inline app/helpers/application_helper.rb]: Fix tabbing and spacing on these lines. Too much content on one line, hurts readability.

File: app/controllers/api/v1/notifications_controller.rb [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+  def index
+    if current_user
+      @notifications = current_user.notifications
+      @posts = Post.all
+      if @notifications
+        data = {
+            notifications: current_user.notifications,
+            announcements: @posts
+        }
+        render json: { status: 'SUCCESS', message: "Notifications for user with netid #{current_user.netid} found.", data: data } , status: :ok
```
Comment (amanmibra) [inline app/controllers/api/v1/notifications_controller.rb:30]: Too much content on one line. On this line, and all the other renders as well. Refer to the render json's in other controllers.

**Review by amanmibra — CHANGES_REQUESTED** [review]
Nice work, bro! I went ahead and went skimmed through your code, and made some comments, check them out. I will make another review once you make more changes.

My experience with creating notifications is good. Notifications with shifts seem to work like they are meant to, but I did not get any announcement notifications when I made a post. Look into why thats happening.

Looks like you are heading in the right direction. Otherwise, all I have left are some notes and thoughts below that didn't really have any space in the code to add a comment to. 

Notes:
- See if you can get the domain that shows up as gthc.io and not as mygthcpush.os.tc. (Also see if a logo can be added but not super important, just find out but dont implement if you feel like it may take you long).
- Add an attribute for the user (that can be added to the settings page) that allows them indicate whether or not they want notifications, and only create notifications based on that condition.
- I do not like the changes that Mike added to our dependencies and Dockerfile and they are causing some breaking issues. I am not completely certain on what to do yet, but it may look like we either remove all of Mike's changes by hand, or we move your changes to a separate branch. 
- Make sure on OneSignal you have a project for prod, staging, and local dev. We cannot have Onesignal data being shared across all environments.

**Comment by amanmibra** [comment]
@shamystic Also question, why do you ask to not make more than one LM Post? Is it so it does not spam to everyone or is there a functionality issue? 

I did not try it more than once as I had issues with getting a notification on the first try like I mentioned above. ^^

File: README.md (hunk 3 above)
Comment (shamystic) [reply README.md]: Done

File: app/controllers/pages_controller.rb (hunk 4 above)
Comment (shamystic) [reply app/controllers/pages_controller.rb]: Not sure why this is here. Will take it out

File: app/models/user.rb (hunk 5 above)
Comment (shamystic) [reply app/models/user.rb]: Done

File: app/views/pages/notif.html.erb (hunk 6 above)
Comment (shamystic) [reply app/views/pages/notif.html.erb]: Will take out the console.log, but otherwise since I have yet to do it via React, I will remove it and the route from routes.rb once we have it fully working via React, as otherwise there is no other way to accept receiving notifications.

File: db/migrate/20200103010349_change_notification_id_to_be_string_in_notifications.rb (hunk 7 above)
Comment (shamystic) [reply db/migrate/20200103010349_change_notification_id_to_be_string_in_notifications.rb:1]: Oh sick will keep that in mind for the future. I'll ask you about if there's a way to clean up the current ones too since it would make things much more organized.

File: public/OneSignalSDKUpdaterWorker.js (hunk 8 above)
Comment (shamystic) [reply public/OneSignalSDKUpdaterWorker.js]: Done

File: app/javascript/packs/hello_react.jsx (hunk 10 above)
Comment (shamystic) [reply app/javascript/packs/hello_react.jsx]: Done

File: app/helpers/application_helper.rb (hunk 12 above)
Comment (shamystic) [reply app/helpers/application_helper.rb]: We can easily add more apps to our OneSignal account. We can add a production app.

File: app/views/pages/notif.html.erb (hunk 6 above)
Comment (amanmibra) [reply app/views/pages/notif.html.erb]: sounds fair, if you push your changes, it will resolve them on github btw

File: app/controllers/pages_controller.rb (hunk 4 above)
Comment (shamystic) [reply app/controllers/pages_controller.rb:18]: Currently this is necessary in order to access the "current_user" attribute for the logged in user for the /notif route. Once I work with you to figure out the proper way to use that code snippet in react, we can clean this up.

File: app/helpers/application_helper.rb (hunk 13 above)
Comment (shamystic) [reply app/helpers/application_helper.rb]: Fixed tabs I think

File: app/controllers/api/v1/notifications_controller.rb (hunk 14 above)
Comment (shamystic) [reply app/controllers/api/v1/notifications_controller.rb:30]: Do you mean that I should not be showing the netid in the message?

File: app/helpers/application_helper.rb (hunk 11 above)
Comment (shamystic) [reply app/helpers/application_helper.rb]: So sorry for this rookie mistake. Fixed (except for the js app id as I need to ask you how to load env variables into the js). Sent you the updated .env file.

File: app/helpers/application_helper.rb (hunk 13 above)
Comment (shamystic) [reply app/helpers/application_helper.rb]: Actually my IDE is being weird and just visually showing me changes in my editor but it's not affecting the actual file. Will get this fixed.

File: README.md [hunk 15]
```diff
@@ -51,24 +51,34 @@ Things you may want to cover:
...
+    
+    b. Always run this command from here on out to start the local server.
 
 ### Local Software Versioning
 It is required that your system is running `Rails 5.0+` for the local development to function properly. It is recommended to use the most recent version of Node, but anything above `Node 8.x` shall work. 
 
 ## Configuration
-__Important:__ There are private environment variables needed for certain aspects of the application. Therefore, it will be required that you have a `.env` file in the root of your project __before__ running anything. Please contact the project leaders (Aman or Anesu) for this information. __Finally, the .env file should not be pushed to github under no circumstances!__
+__Important:__ There are private environment variables needed for certain aspects of the application. Therefore, it will be required that you have a `.env` file in the root of your project __before__ running anything. Please contact the project leaders (Aman or Anesu) for this information. __Finally, the .env file should not be pushed to github under no circumstances!__
```
Comment (Akhan221) [inline README.md:84]: I know you provided the updated .env file in the dev chat, but maybe mention that an updated .env file is required in the readme?

File: app/helpers/application_helper.rb [hunk 16]
```diff
@@ -1,2 +1,109 @@
...
+
+	def shift_notification(shift, title: nil, content: nil, test: false, min_before: 10, send_now: false)
+		# Send to all shift members
+		netids = shift.users.collect(&:netid)
+		# Notification time is min_before minutes before shift, if not sending now
+		time = send_now ? nil : shift.start_time - min_before*60
+		title = "Upcoming Shift Reminder" if title.nil?
+		if content.nil?
+			content = "Hey! Your tent shift in K-Ville starts in #{min_before} minutes!"
+		end
```
Comment (Akhan221) [inline app/helpers/application_helper.rb:17]: I like that you kept the message dynamic with {min_before}, later down the road users can possibly adjust when they receive their notification.

**Review by Akhan221 — COMMENTED** [review]
Great work putting this together Shamikh! Aman seemed to have caught most of the errors in your code so I will delve deeper into the functionality of the notifications. 

1. I was able to successfully allow notifications and after creating a shift, I received the proper notification. Updating the shift and adding users works properly as well. As Vinit and I tested, if a user has signed in on mobile and desktop, both machines receive the notification. 
Note: mobile notifications are sometimes showing up only when a user opens the app, not in the notification center. That may be due to user's settings, further testing may be needed. 

2. Once line monitor posts are up and running testing will be done.

3. Testing Notifications Endpoint is working properly, kept track after updating and creating shifts.
